server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -924,6 +1015,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
     }
   }
 
+  private boolean needsFencingForDeletion(MergeInfo info, KeyExtent keyExtent) {
+    return needsFencingForSplit(info, keyExtent) || info.deleteOverlaps(keyExtent);
+  }
```

Comment:
This is the only method that calls needsFencingForSplit. needsFencingForSplit also double-checks that the operation `isDelete()`, but that's redundant if this method is being called.

Putting the contents of that method into this one, and simplifying it dramatically, I get:

```suggestion
  private boolean needsFencingForDeletion(MergeInfo info, KeyExtent keyExtent) {
    // Does this extent cover the end points of the delete?
    Predicate<Text> isWithin = r -> r != null && keyExtent.contains(r);
    Predicate<Text> isNotBoundary =
        r -> !r.equals(keyExtent.endRow()) && !r.equals(keyExtent.prevEndRow());
    KeyExtent deleteRange = info.getExtent();
    return (keyExtent.overlaps(deleteRange) && Stream
        .of(deleteRange.prevEndRow(), deleteRange.endRow()).anyMatch(isWithin.and(isNotBoundary)))
        || info.deleteOverlaps(keyExtent);
  }
```

Now, I'm still not exactly sure about the edge cases this is checking for... there are too many overlaps, contains, and boundary checks happening in different methods to really intuitively get it. But this simplification should help. I selected predicate names that suggest what is being checked here, so that it would be easier to understand why you might need fencing in some cases, but not in others. You don't need fencing, for example, if the delete range's endRow is within the current keyExtent, but falls exactly on its boundary... you can just delete all the files. But you would need to fence if the delete range is within the current keyExtent, but doesn't reach all the way at the boundary.

I'm really not sure about the correctness of this when it comes to the edge cases, though, like the first or last tablet (where endRow or prevEndRow are null), or if the bounds of the deleted range are not inclusive. I'm hoping there's already good coverage of all this, so you don't have to do much... but simplification like what I am suggesting here could make it much more readable if done in more places.

Reply:
Thanks I will take a look when I get a chance and try and see if it's correct and the changes make sense. It's tricky as the logic combines checks for both when we need to do chop compactions and also if we need to split in order to see if we need to fence the files instead. Also once things are working and we have good test coverage we can always then clean up or improve the code and business logic. I figure some of this will be improved a bit over time as we test and find ways to optimize.

Reply:
My latest commit uses this suggestion, it seems to be equivalent after looking at it and is a lot more concise. All the tests still pass.